Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift

```diff
@@ -121,15 +122,21 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
                 forResult: lookupResult,
                 completion: completionHandler
             )
-        case .error, .timeout:
-            let userInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
-            var errorCode: Int = BTThreeDSecureError.unknown.errorCode
+        case .error:
+            let errorUserInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
 
             if validateResponse.errorNumber == 1050 {
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: errorUserInfo))
             }
             apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
-            completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
+        case .timeout:
+            let timeoutUserInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+
+            if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                completionHandler(nil, NSError(domain: timeoutUserInfo.description, code: errorCode, userInfo: timeoutUserInfo))
```

Comment:
```suggestion
            completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
```

Reply:
I tried running your suggestion and setting a breakpoint at `validateResponse.actionCode`. The `errorCode` returns as `0` instead of `9` when I reach the max timeout limited

Reply:
`9` is the error code for `.exceededTimoutLimit` and it's what was missing before. I know that, in theory, the suggested refactored code should work. I'm using `po errorCode` in the debugger but might need to test something else that returns the correct error code and error description, so let me know